uqcsbot/course_ecp.py

```diff
@@ -0,0 +1,115 @@
+    Offering,
+    HttpException,
+    CourseNotFoundException,
+    ProfileNotFoundException,
+    get_course_profile_url,
+)
+from uqcsbot.yelling import yelling_exemptor
+
+class CourseECP(commands.Cog):
+    def __init__(self, bot: commands.Bot):
+        self.bot = bot
+
+    @app_commands.command()
+    @app_commands.describe(
+        course1="The course to find an ECP for.",
+        course2="The second course to find an ECP for.",
+        course3="The third course to find an ECP for.",
+        course4="The fourth course to find an ECP for.",
+        year="The year to find the course ECP for. Defaults to what UQCSbot believes is the current semester.",
+        semester="The semester to find the course ECP for. Defaults to what UQCSbot believes is the current semester.",
+        campus="The campus the course is held at. Defaults to St Lucia. Defaults to St Lucia. Note that many external courses are 'hosted' at St Lucia.",
+        mode="The mode of the course. Defaults to Internal.",
+    )
+    @yelling_exemptor(
+        input_args=["course1, course2, course3, course4"]
+    )    
+    
+    async def courseecp(
+        self,
+        interaction: discord.Interaction,
+        course1: str,
+        course2: Optional[str],
+        course3: Optional[str],
+        course4: Optional[str],
+        year: Optional[int] = None,
+        semester: Optional[Offering.SemesterType] = None,
+        campus: Offering.CampusType = "St Lucia",
+        mode: Offering.ModeType = "Internal",
+    ):
+        """
+        Returns the URL of the ECPs for course codes given. Assumes the same semester and year for the course codes given.
+
+        """
+        await interaction.response.defer(thinking=True)
+
+        possible_courses = [course1, course2, course3, course4]
+        course_names = [c.upper() for c in possible_courses if c != None]
+        course_name_urls: dict[str,str] = {}
+        offering = Offering(semester=semester, campus=campus, mode=mode)
+        
+        try:
+            for course in course_names:
+                course_name_urls.update({course: get_course_profile_url(course, offering, year)})
+        except HttpException as exception:
+            logging.warning(
+                f"Received a HTTP response code {exception.status_code}. Error information: {exception.message}"
+            )            
+            await interaction.edit_original_response(
+                content=f"Could not contact UQ, please try again."
+            )
+            return
+        except (CourseNotFoundException, ProfileNotFoundException) as exception:     
+            await interaction.edit_original_response(content=exception.message)
+            return
+        
+        # If year is none assign it the current year
+        if not year:
+            year = datetime.today().year
+        
+        # If semester is none assign it the current estimated semester
+        if not semester:
```

Comment:
This is a sign that the `Offering` code is not great (not your code), but it would be good to use the `_estimate_current_semester()` method. I'd potentially make `_estimate_current_semester()` a non-private function within `uq_course_utils.py`. Note that if you do this, you'll have to slightly change the constructor for `Offering`.

Reply:
Agreed, however I don't know where to start to change the constructor for Offering, any ideas?

Reply:
Going to attempt to make the function not static but keep it in `Offering`, it makes more sense to me in how the rest of the functions in `Offering` are called in `past_exams.py` and `whatsdue.py`. Hopefully making this change doesn't break anything

Reply:
huh don't worry, doesn't need to be not static, it just needs to be unprotected